### PR TITLE
Added binary numeric conversions for ternary types

### PR DESCRIPTION
**Pull request by zesterer**
# Description of change

Added binary integer <-> ternary conversion

## Links to any relevant issues

https://github.com/iotaledger/bee-rfcs/pull/37

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Tests included in the PR.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: bee-ternary/Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,7 @@ homepage = "https://www.iota.org"
 byteorder = "1.3.4"
 lazy_static = "1.4.0"
 serde = { version = "1.0", optional = true }
+num-traits = "0.2"
```
Comment (thibault-martinez) [inline bee-ternary/Cargo.toml:17]: Do you mind ordering the deps alphabetically :) ?

File: bee-ternary/src/convert.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+fn trits_to_int<I, T: RawEncoding<Trit = Btrit> + ?Sized>(trits: &Trits<T>) -> Result<I, Error>
+where
+    I: Copy + FromPrimitive + ToPrimitive + Bounded + Zero + One + Neg<Output = I> + Add + Mul,
+{
+    if trits.is_empty() {
+        Err(Error::Empty)
+    } else {
+        let max_trits: usize = max_trits::<I>();
+
+        assert!(max_trits == 41, "{}", max_trits);
```
Comment (thibault-martinez) [inline bee-ternary/src/convert.rs:63]: I'm not sure to understand this assert

File: bee-ternary/src/convert.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+        if trits.len() > max_trits {
+            // TODO test
```
Comment (thibault-martinez) [inline bee-ternary/src/convert.rs:66]: Can we actually test both scenario ? Shouldn't be that hard.
- A buffer with more than max_trits trits but all trits afterthat amount are 0 : should pass
- A buffer with more than max_trits trits and at least one trits after that amount is 1 : should not pass

File: bee-ternary/src/convert.rs (hunk 2 above)
Comment (zesterer) [reply bee-ternary/src/convert.rs:63]: Whoops, that should have been gotten rid of. It's there as a test.

File: bee-ternary/src/convert.rs (hunk 3 above)
Comment (zesterer) [reply bee-ternary/src/convert.rs]: Just added a test for this

File: bee-ternary/Cargo.toml (hunk 1 above)
Comment (zesterer) [reply bee-ternary/Cargo.toml:17]: Sure, will do this

**Comment by thibault-martinez** [comment]
Can you prepare an entry in the CHANGELOG for 0.2.0-alpha please ?

File: bee-ternary/src/trit/mod.rs [hunk 4]
```diff
@@ -11,23 +11,36 @@
...
     fn zero() -> Self;
 
+    /// Turn this trit reference into one with an arbitrary lifetime.
+    ///
+    /// Note that this is largely an implementation detail and is rarely useful for API users.
     fn as_arbitrary_ref<'a>(&self) -> &'a Self;
 }
 
+/// A trait implemented by trits that can be shifted between balance domains.
+// TODO: Is this a good API?
```
Comment (thibault-martinez) [inline bee-ternary/src/trit/mod.rs:39]: What's the issue with it ?

File: bee-ternary/src/lib.rs [hunk 5]
```diff
@@ -734,6 +764,57 @@ impl<T: RawEncodingBuf> TritBuf<T> {
...
+    pub fn pad_zeros(&mut self) {
+        while self.len() % 3 != 0 {
+            self.push(Btrit::Zero);
+        }
+    }
+
+    /// Pad the trit buffer with [`Btrit::Zero`] until the buffer's length is a multiple of 3.
+    ///
+    /// This method is often used in conjunction with [`Trites::as_trytes`].
+    pub fn padded_zeros(mut self) -> Self {
```
Comment (thibault-martinez) [inline bee-ternary/src/lib.rs:780]: Why do we need both ?

File: bee-ternary/Cargo.toml [hunk 6]
```diff
@@ -13,6 +13,7 @@ homepage = "https://www.iota.org"
 [dependencies]
 byteorder = "1.3.4"
 lazy_static = "1.4.0"
```
Comment (thibault-martinez) [inline bee-ternary/Cargo.toml:15]: I forgot to remove that when moving bigint, can you remove it please :( ?

File: bee-ternary/Cargo.toml [hunk 7]
```diff
@@ -13,6 +13,7 @@ homepage = "https://www.iota.org"
 [dependencies]
 byteorder = "1.3.4"
```
Comment (thibault-martinez) [inline bee-ternary/Cargo.toml:14]: Well this one too...

File: bee-ternary/CHANGELOG.md [hunk 8]
```diff
@@ -19,6 +19,24 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Security -->
 
+## 0.2.0-alpha - 2020-07-D16
```
Comment (thibault-martinez) [inline bee-ternary/CHANGELOG.md:22]: D ?

File: bee-ternary/CHANGELOG.md [hunk 9]
```diff
@@ -19,6 +19,24 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
+### Changed
+
+### Deprecated
+
+### Removed
+
+### Fixed
+
+### Security
```
Comment (thibault-martinez) [inline bee-ternary/CHANGELOG.md:38]: Can we remove sections is empty ? Like the previous release.

File: bee-ternary/src/lib.rs (hunk 5 above)
Comment (zesterer) [reply bee-ternary/src/lib.rs:780]: I imagine this operation would be common immediately after creation so this allows it to be used inline

File: bee-ternary/src/trit/mod.rs (hunk 4 above)
Comment (zesterer) [reply bee-ternary/src/trit/mod.rs:39]: It feels bizarre for the same method to shift both ways. `.into_balance()` / `.into_unbalanced()` seems like the more natural API to me.

Comment (thibault-martinez) [reply bee-ternary/src/trit/mod.rs:39]: But then you have to know the concrete type ? You can't "toggle" it anymore ?

File: bee-ternary/src/convert.rs [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+// TODO: Find a way to implement this without conflicting impls
+// impl<I, T: RawEncodingBuf> From<I> for TritBuf<T>
+// where
+//     T::Slice: RawEncoding<Trit = Btrit>,
+//     I: AsPrimitive<i8> + FromPrimitive + Signed,
+// {
+//     fn from(x: I) -> Self {
+//         signed_int_trits(x).collect()
+//     }
+// }
```
Comment (thibault-martinez) [inline bee-ternary/src/convert.rs:35]: Is this still relevant ?

File: bee-ternary/src/convert.rs [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+            let modulus = (x % radix).as_();
+            x = x / radix;
+            Some(Utrit::try_from(modulus).unwrap())
+        }
+    })
+    // If the integer is exactly 0, add an extra trit
+    .chain(Some(Utrit::Zero).filter(|_| x.is_zero()))
+}
+
+#[cfg(test)]
```
Comment (thibault-martinez) [inline bee-ternary/src/convert.rs:197]: Should the tests be in `tests` for consistency ?

Comment (zesterer) [reply bee-ternary/src/convert.rs:197]: Probably, yes. Will change this.

File: bee-ternary/src/convert.rs (hunk 10 above)
Comment (zesterer) [reply bee-ternary/src/convert.rs:35]: I'd rather keep for future reference it since it's the 'preferred' way of doing things.

Comment (thibault-martinez) [reply bee-ternary/src/convert.rs:35]: Sure

File: bee-ternary/src/trit/mod.rs (hunk 4 above)
Comment (zesterer) [reply bee-ternary/src/trit/mod.rs:39]: Sure, I'm just not convinced that's always what is needed.

**Pull request merged**